### ISSUE TITLE
Seacas/STK: How to save (and re-load later) fields that are not transient in Exodus mesh

### ISSUE THREAD

**Issue by bartgol** [trilinos/Trilinos#5371]
## Question
@trilinos/stk @trilinos/seacas 

I realized that the STK-Exodus combo does not let me store/load arbitrary fields with role type `MESH`. It seems that only a handful of pre-defined names can be correctly handles as `MESH` fields (e.g., "coordinates", "ids", "connectivity",...). It seems that the culprit is how exodos treats stored variables, namely always associating them with a time stamp (save for the coordinates of the mesh).

My question is: how can I store/load a field that is "intrinsically" not a transient field, without doing it at every single time step? I am assuming that a `TRANSIENT` field is written in the mesh every time, which would be a waste of storage for a steady field...

Edit: to add more context, I ran into this issue when calling the `field_data_from_ioss` function from `IossBridge.*pp` in stk_io. This leads to a bunch of calls in Ioss, which ultimately lead to calling `get_field_internal` in one of the concrete classes derived from `Ioss::DatabaseIO` (e.g., in my case, `Iofx::DatabaseIO`). This function checks the field role, and only allows a limited set of names for the role `MESH` (as noted above), throwing a warning if another name is encountered. I worked my way around this by marking all my fields as `TRANSIENT`. This however, forced me to manually call the `begin_state` method on the underlying input `Ioss::Region`, or else I would get the error

```
 ERROR: No currently active state.  The calling code must call Ioss::Region::begin_state(int step)
        to set the database timestep from which to read the transient data.
        [ebs-steady.exo]
```

Besides the complication of having steady fields written multiple times to file, this seems overly complicated, since it forces me to go into the details of not only stk, but also Ioss. Ideally, one would like to interact only with the `StkMeshIoBroker` class (after all, that's what brokers are for).

Am I missing something here? Or is this a feature lacking in stk/seacas, which cannot possibly be accommodated (due to upstream library limitations)?

**Comment by mhoemmen** [comment]
@trilinos/seacas @gsjaardema

**Comment by gsjaardema** [comment]
Maybe attributes would do what you need.  Role would be `ATTRIBUTE`. They can be defined on a block-by-block or set-by-set basis and are named.

**Comment by bartgol** [comment]
I thought attributes were single scalars defined for the whole [Elem|Node|Face|Edge]Block, rather than real fields defined on each entity...

**Comment by gsjaardema** [comment]
No, an attributed is a per-entity value.  For example, shells typically have a `THICKNESS` attribute which gives the value of the shell thickness for each element in the element block.

**Comment by bartgol** [comment]
Thanks @gsjaardema ! I will try this approach and see if it's what I need (it sounds like, from what you said). I'm closing this for now.

**Comment by bartgol** [comment]
@gsjaardema I am reopening this issue, since I have a followup question that fits in this discussion: I have tried to add an attribute field, to export mesh-related information in the exodus file. I tried this, during the 'setup' of the IO broker:
```

    auto& my_mesh_field = ...;
    stk::io::OutputVariableParams vars(my_mesh_field.name());
    mesh_data->add_attribute_field(outputFileIdx, my_mesh_field, vars);
```

The 'write' phase happens with the sequence

```
mesh_data->begin_output_step(outputFileIdx, timeLabel);
mesh_data->write_defined_output_fields(outputFileIdx);
// a for loop with several "mesh_data->write_global(...);" 
mesh_data->end_output_step();
```

However, I don't see the field in the exodus file. Am I missing some step for guaranteeing the mesh attribute field is added?

Thanks!

**Comment by bartgol** [comment]
Nevermind, I think it did get exported, it's just that ParaView does not display mesh fields (is that right?). I used exotxt to inspect the exodus file, and I found one attribute, although the values are all wrong (it contains e-314 values, while my field was integer).

Are attribute fields perhaps not the best approach to save mesh-related information in a way that can be visualized with tools like Paraview? Or is there a way to visualize mesh attribute fields in Paraview?
